### PR TITLE
[Feature] support mtp3

### PR DESCRIPTION
**Pull request by mingMelody**
## Purpose
To support MTP3, this PR introduces the **MultiLayerEagleProposer** class. The main challenges of MTP3 are illustrated in the figure below.

image

During multi-layer draft model execution, a naive just-roll strategy leads to incorrect KV cache states. In such scenarios, the affected tokens must be recomputed. However, direct recomputation depends on the hidden states produced in the previous iteration, which would require storing hidden states from (layer_num − 1) layers and handling a large number of corner cases.

To address this, this PR introduces an `adjust_input` function that performs input shifting before entering the MTP-layer inference. This proactively masks potential corner cases that could arise in future steps. As a result, the inference phase only needs to perform a single roll operation to proceed correctly.

This approach applies one-time handling at boundary conditions and only caches the target model’s hidden states, leading to a simpler overall design with minimal additional overhead.

## Test Plan
Need to run with `enable_multi_layers_mtp` in `speculative_config` to enable MTP3.

A example of `step3p5-flash` with mtp3 as below:
```
vllm serve <MODEL_PATH_OR_HF_ID> \
  --served-model-name step3p5-flash \
  --tensor-parallel-size 8 \
  --enable-expert-parallel \
  --disable-cascade-attn \
  --reasoning-parser step3p5 \
  --enable-auto-tool-choice \
  --tool-call-parser step3p5 \
  --speculative_config '{"method": "step3p5_mtp", "num_speculative_tokens": 3, "enable_multi_layers_mtp": true}' \
  --trust-remote-code 
```

### acc tests
```
# ACC Test
python examples/offline_inference/spec_decode.py --num-prompts 80 --dataset-name hf --dataset-path philschmid/mt-bench --method mtp --model-dir <MODEL_PATH_OR_HF_ID> --num-spec-tokens 3 --tp 8 --enable-multi-layers-mtp

# Test Result
--------------------------------------------------
total_num_output_tokens: 20207
num_drafts: 7176
num_draft_tokens: 21528
num_accepted_tokens: 13042
mean acceptance length: 2.82
--------------------------------------------------
acceptance at token 0: 0.80
acceptance at token 1: 0.59
acceptance at token 2: 0.42
```

### unit tests
```
# Run unit tests
pytest tests/v1/spec_decode/test_mtp3.py -v 

# Test Result
============================= test session starts ==============================
platform linux -- Python 3.12.12, pytest-8.3.5, pluggy-1.6.0 -- /home/i-zhangmingming/open_source/vllm/.venv/bin/python
cachedir: .pytest_cache
rootdir: /home/i-zhangmingming/open_source/vllm
configfile: pyproject.toml
plugins: anyio-4.12.1
collecting ... collected 12 items

tests/v1/spec_decode/test_mtp3.py::test_adjust_input_layer3_cases[shift_0_at_sequence_end] PASSED [  8%]
tests/v1/spec_decode/test_mtp3.py::test_adjust_input_layer3_cases[batch2_short_seq_no_shift] PASSED [ 16%]
tests/v1/spec_decode/test_mtp3.py::test_adjust_input_layer3_cases[batch2_short_seq_shift_on_first] PASSED [ 25%]
tests/v1/spec_decode/test_mtp3.py::test_adjust_input_layer3_cases[short_seq_len_2_shift_0_cache_len_1] PASSED [ 33%]
tests/v1/spec_decode/test_mtp3.py::test_adjust_input_layer3_cases[short_seq_len_2_shift_1_cache_len_2] PASSED [ 41%]
tests/v1/spec_decode/test_mtp3.py::test_adjust_input_layer3_cases[shift_bounded_by_start_pos_zero] PASSED [ 50%]
tests/v1/spec_decode/test_mtp3.py::test_adjust_input_layer3_cases[shift_bounded_by_start_pos] PASSED [ 58%]
tests/v1/spec_decode/test_mtp3.py::test_adjust_input_layer3_cases[shift_2_bounded_by_remaining] PASSED [ 66%]
tests/v1/spec_decode/test_mtp3.py::test_adjust_input_layer3_cases[shift_3_full_cache_window] PASSED [ 75%]
tests/v1/spec_decode/test_mtp3.py::test_adjust_input_layer3_cases[batch2_shift_1_and_1] PASSED [ 83%]
tests/v1/spec_decode/test_mtp3.py::test_adjust_input_layer3_cases[batch4_mixed_shifts] PASSED [ 91%]
tests/v1/spec_decode/test_mtp3.py::test_adjust_input_layer3_cases[batch2_shift_0_and_2] PASSED [100%]
======================= 12 passed, 2 warnings in 11.45s ========================
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/v1/core/kv_cache_utils.py [hunk 1]
```diff
@@ -1050,19 +1052,28 @@ def _get_kv_cache_groups_uniform_page_size(
...
-        # instead of layers[i * group_size: (i + 1) * group_size]
-        for i in range(num_groups):
-            grouped_layers.append(layers[i::num_groups])
+        # for support multi layer mtp, we need to
+        # make all mtp layers in the same group
+        if (
+            vllm_config.speculative_config is not None
+            and vllm_config.speculative_config.enable_multi_layers_mtp
+        ):
+            for i in range(0, len(layers), group_size):
```
Comment (benchislett) [inline vllm/v1/core/kv_cache_utils.py:1073]: Is this the right way to handle this? Can any vLLM KV-cache-manager experts weigh in here?

File: vllm/v1/spec_decode/eagle.py [hunk 2]
```diff
@@ -381,8 +381,11 @@ def propose(
             input_ids = None
             inputs_embeds = self.inputs_embeds[:num_input_tokens]
         else:
+            self.inputs_embeds[:num_tokens] = self.model.embed_input_ids(
+                self.input_ids[:num_tokens]
+            )
             input_ids = self.input_ids[:num_input_tokens]
-            inputs_embeds = None
+            inputs_embeds = self.inputs_embeds[:num_input_tokens]
```
Comment (benchislett) [inline vllm/v1/spec_decode/eagle.py]: Why is this change necessary?

File: vllm/v1/spec_decode/multi_layer_eagle.py [hunk 3]
```diff
@@ -0,0 +1,925 @@
...
+                "does not match layer_num, adjusting to layer_num"
+            )
+            self.num_speculative_tokens = self.layer_num
+        self.running_req_ids: list[str] | None = None
+        self.draft_input_states_pool: dict[str, DraftInputStates] = {}
+
+    def set_running_req_ids(self, req_ids: list[str]):
+        self.running_req_ids = req_ids
+
+    def _get_draft_input_states(self, req_id: str, len: int) -> DraftInputStates:
```
Comment (benchislett) [inline vllm/v1/spec_decode/multi_layer_eagle.py]: Is this method unused?

File: vllm/v1/spec_decode/multi_layer_eagle.py [hunk 4]
```diff
@@ -0,0 +1,925 @@
...
+        assert isinstance(out_cached_positions, list)
+        for row in range(3):
+            _shift_and_gather_cache_1d_kernel(
+                src_positions[row],
+                dst_positions[row],
+                cached_prev_positions[row],
+                out_cached_positions[row],
+            )
+
+    _ml_eagle_shift_hidden_kernel[(batch_size, num_blocks, num_hidden_blocks)](
```
Comment (benchislett) [inline vllm/v1/spec_decode/multi_layer_eagle.py]: what is the significance of the `_ml` prefix?

**Review by benchislett — CHANGES_REQUESTED** [review]
I have a few high-level concerns with this implementation:

### Correctness details
There is a fundamental train/inference mismatch for MTP modules, so it is hard to define what the ideal "correct" behaviour should be. There is some room for interpretation. That said, this PR seems to overlook a detail which is central to a lot of discussion around multi-head MTP implementations:

When recomputing with only as many cached hidden states as rejected tokens, the input shapes are consistent but the KV cache of the later MTP modules is not consistent. Your example figure correctly highlights that when rejecting tokens and naively rolling, the KV cache of the later MTP modules are corrupted with bad tokens and hidden states. However, when rolling based on the number of accepted tokens, the later modules' KV caches are populated based on the hidden states of the previous MTP iterations, and not the hidden states from the base model. Consider the case where all 3 tokens are accepted, and therefore no adjusting is necessary: the input hidden states to MTP0 are "7,8,9,10" with the corresponding hidden states. Correspondingly the inputs to MTP2 are "9,10,11,12". In this case the KV cache for MTP2 for tokens "7,8" are the hidden states _from MTP1_ for those tokens, not the new target model hidden states from the verification of those tokens. This creates an inconsistency for MTP1 and MTP2 where their context states are derived from a mix of target and MTP output hidden states.

It is debatable whether this is even worth considering as long as the token ids are correct, since MTP are typically trained based on the inputs of the previous modules as context anyways, but nevertheless it is an aspect of correctness worth considering. The 'corrected' solution in this case would be to prepend the cached tokens/hiddens always, ensuring that MTP2 always updates its KV cache with the target model's hidden states. This does mean that the input shapes would increase and no longer be the same as the shapes from the target model. In this case, reusing some logic from [Parallel Drafting](https://github.com/vllm-project/vllm/pull/32887) (which similarly has to insert tokens into the batch for specdec) would be useful.

### Duplicated Code
The MultiLayerEagleProposer adds a lot of new code to maintain for EAGLE speculative decoding. I would hope that this can be implemented similar to Draft Models and Parallel Drafting where most of the EAGLE inference code is reused, for maintainability.

### Caching state on the drafter
It is a major challenge to maintain consistent state in the GPU Model Runner. The input_batch class has a lot of utilities to ensure that under a dynamically changing batch, the state tensors remain consistent and input preparation is efficient. I am opposed to the style of decomposing the state into individual tensors and managing them in the Proposer class. I feel this will lead to a lot of subtle bugs when the batch is reordered in unexpected ways, and/or a lot of overhead needing to rebuild the batch from scratch every iteration. There are a lot of clone/copy/insert operations that look like they would cause a slowdown for large batch sizes. Have you measured any overheads of this approach compared to EAGLE-style MTP, across a range of batch sizes? Does the evaluation and drafting accuracy remain stable across batch sizes?

### Underdocumented kernels
This PR introduces several custom triton kernels for preparing the metadata. Being less readable than straight pytorch code, they should be documented with comments explaining their purpose and input/outputs

I am happy to discuss in more detail offline in the vLLM slack if desired. Feel free to reach out directly or in #feat-spec-decode.

File: vllm/v1/spec_decode/eagle.py (hunk 2 above)
Comment (mingMelody) [reply vllm/v1/spec_decode/eagle.py]: Not necessary, have removed it.

File: vllm/v1/spec_decode/multi_layer_eagle.py (hunk 4 above)
Comment (mingMelody) [reply vllm/v1/spec_decode/multi_layer_eagle.py]: remove it.

File: vllm/v1/spec_decode/multi_layer_eagle.py (hunk 3 above)
Comment (mingMelody) [reply vllm/v1/spec_decode/multi_layer_eagle.py]: I will try to move the cache logic into req_states. After that, it will no longer be necessary.

File: vllm/v1/spec_decode/eagle.py [hunk 5]
```diff
@@ -79,6 +79,9 @@ def __init__(
         self.dp_rank = vllm_config.parallel_config.data_parallel_rank
         self.num_speculative_tokens = self.speculative_config.num_speculative_tokens
 
+        self.enable_multi_layers_mtp = self.speculative_config.enable_multi_layers_mtp
+        self.layer_num = 1
```
Comment (benchislett) [inline vllm/v1/spec_decode/eagle.py:82]: This needs a better name, or a comment explaining what "layer_num" means. Intuitively, I would assume it means "number of layers in each MTP/EAGLE module", but that seems incorrect.

File: vllm/v1/spec_decode/multi_layer_eagle.py [hunk 6]
```diff
@@ -0,0 +1,555 @@
...
+            start_token_pos,
+        )
+        shift = torch.clamp(shift, min=0)
+
+        # Metadata updates (matches the original reference implementation).
+        token_indices_to_sample.add_(shift)
+        common_attn_metadata.seq_lens.sub_(shift)
+
+        # NOTE: ignore cpu data to avoid device sync
+        # common_attn_metadata.seq_lens_cpu.copy_(common_attn_metadata.seq_lens,
```
Comment (benchislett) [inline vllm/v1/spec_decode/multi_layer_eagle.py]: Is this dead code?

File: vllm/v1/spec_decode/multi_layer_eagle.py [hunk 7]
```diff
@@ -0,0 +1,555 @@
...
+
+        return prev_token_ids, prev_positions, prev_hidden_states, common_attn_metadata
+
+    def prepare_inputs(
+        self,
+        common_attn_metadata: CommonAttentionMetadata,
+        sampled_token_ids: list[list[int]],
+        num_draft_tokens: list[int],
+    ) -> tuple[CommonAttentionMetadata, torch.Tensor]:
+        """
```
Comment (benchislett) [inline vllm/v1/spec_decode/multi_layer_eagle.py]: Doc comment not needed if this function raises an error anyways.

File: vllm/v1/spec_decode/multi_layer_eagle.py [hunk 8]
```diff
@@ -0,0 +1,555 @@
...
+        This function is used to prepare the inputs for speculative decoding.
+        It updates to the common_attn_metadata to account for the rejected
+        tokens (and newly sampled tokens). It also returns the token indices
+        of the tokens that should be fed to the speculator.
+        """
+        raise Exception(
```
Comment (benchislett) [inline vllm/v1/spec_decode/multi_layer_eagle.py]: Generic exception is not recommended.

I suggest you ensure that when using multi layer eagle the padded drafter batch mode be automatically enabled as required ([example](https://github.com/vllm-project/vllm/pull/36847/changes#diff-d9f90bb1c6d4f572294cec8ce7251c46224d8f6ecf5918d7f8a9fb768bb65dbbR552)). If so, you can change this to an `assert`

File: vllm/v1/spec_decode/multi_layer_eagle.py [hunk 9]
```diff
@@ -0,0 +1,555 @@
...
+                    "positions": self._get_positions(num_input_tokens),
+                    "hidden_states": self.hidden_states[:num_input_tokens],
+                    "inputs_embeds": inputs_embeds,
+                    "spec_step_idx": fwd_idx,
+                }
+
+                self.model(**model_kwargs)
+
+
+def _multi_layer_eagle_shift_and_cache(
```
Comment (benchislett) [inline vllm/v1/spec_decode/multi_layer_eagle.py:212]: Please add a detailed comment here. It is unclear what the responsibility of this function is

File: vllm/v1/spec_decode/multi_layer_eagle.py [hunk 10]
```diff
@@ -0,0 +1,555 @@
...
+        (
+            common_attn_metadata.query_start_loc_cpu[1:]
+            - common_attn_metadata.query_start_loc_cpu[:-1]
+        )
+        .max()
+        .item()
+    )
+    num_blocks = max(1, (max_window_len + BLOCK_TOKENS - 1) // BLOCK_TOKENS)
+
+    _shift_and_gather_cache_1d_kernel[(batch_size, num_blocks)](
```
Comment (benchislett) [inline vllm/v1/spec_decode/multi_layer_eagle.py:273]: This logic seems very involved. Can you explain why such a complicated implementation is necessary?

Comment (benchislett) [reply vllm/v1/spec_decode/multi_layer_eagle.py:273]: Given all the complexity here, a direct torch.compile'd implementation may be preferable

File: vllm/v1/spec_decode/eagle.py [hunk 11]
```diff
@@ -453,53 +489,85 @@ def propose(
         if num_tokens_across_dp is not None:
             num_tokens_across_dp[self.dp_rank] = num_input_tokens
 
-        if self.supports_mm_inputs:
-            mm_embeds, is_mm_embed = mm_embed_inputs or (None, None)
+        draft_token_ids_list = []
+        for spec_step_idx in range(self.layer_num):
```
Comment (benchislett) [inline vllm/v1/spec_decode/eagle.py:476]: Please avoid implementing this feature in such a way that affects the readability of the codebase more broadly. The core flow of the EAGLE pathway should be preserved as much as possible.

File: vllm/v1/spec_decode/eagle.py [hunk 12]
```diff
@@ -400,6 +421,21 @@ def propose(
             )
             assert target_hidden_states.shape[-1] == self.hidden_size
 
+        (
+            target_token_ids,
+            target_positions,
+            target_hidden_states,
+            common_attn_metadata,
+        ) = self.adjust_input(
```
Comment (benchislett) [inline vllm/v1/spec_decode/eagle.py:441]: What is the purpose of this function? Why can't the multi layer component just specialize `set_inputs_first_pass`?

File: vllm/v1/worker/gpu_model_runner.py [hunk 13]
```diff
@@ -1655,6 +1700,17 @@ def _prepare_inputs(
             self.num_decode_draft_tokens.np[num_reqs:].fill(-1)
             self.num_decode_draft_tokens.copy_to_gpu()
 
+        if self.enable_multi_layer_eagle:
+            multi_layer_eagle_metadata = MultiLayerEagleMetadata(
```
Comment (benchislett) [inline vllm/v1/worker/gpu_model_runner.py]: Why is this metadata object required? What specifically is preventing these fields from living exclusively on the drafter? It is not clear to me why this input_batch metadata is needed any why it must be connected to the drafter

**Review by benchislett — COMMENTED** [review]
This is my first round of feedback. Overall, the code currently feels very opaque and a bit bloated. It has been challenging to parse some of the segments, and while I don't doubt the correctness the complexity feels excessive given the scope.

File: vllm/v1/worker/gpu_input_batch.py [hunk 14]
```diff
@@ -53,6 +53,13 @@ class CachedRequestState:
     pooling_params: PoolingParams | None = None
     pooling_states: PoolingStates | None = None
 
+    # for multi layer eagle proposer
+    cached_len: torch.Tensor | None = None
```
Comment (benchislett) [inline vllm/v1/worker/gpu_input_batch.py:57]: I still don't feel great about having to cache all this state on the drafter. This feels like the wrong way to handle it

Comment (benchislett) [reply vllm/v1/worker/gpu_input_batch.py:57]: In the meantime, can you add some documentation here about what these tensors represent, and how they are intended to be used for multi-layer EAGLE statefulness across iterations?

File: vllm/v1/spec_decode/multi_layer_eagle.py (hunk 10 above)
Comment (mingMelody) [reply vllm/v1/spec_decode/multi_layer_eagle.py:273]: I’m not very familiar with `torch.compile`, so I wanted to clarify the expectation here.

Is the idea to rewrite this in terms of regular PyTorch ops and then rely on torch.compile to fuse and optimize them?

File: vllm/v1/spec_decode/eagle.py (hunk 12 above)
Comment (mingMelody) [reply vllm/v1/spec_decode/eagle.py:441]: Good question. I’ll try folding this into `set_inputs_first_pass` instead.

File: vllm/v1/spec_decode/eagle.py (hunk 11 above)
Comment (mingMelody) [reply vllm/v1/spec_decode/eagle.py:476]: One other option here would be to factor the first-pass logic into a separate function and override that in multi_layer_eagle, so the core EAGLE flow stays mostly unchanged.

Do you think that would be a reasonable approach?

File: vllm/v1/worker/gpu_input_batch.py (hunk 14 above)
Comment (mingMelody) [reply vllm/v1/worker/gpu_input_batch.py:57]: I’ll take another pass to confirm what cache information is actually necessary here, and improve the related documentation accordingly.

File: vllm/v1/spec_decode/multi_layer_eagle.py (hunk 10 above)
Comment (benchislett) [reply vllm/v1/spec_decode/multi_layer_eagle.py:273]: Yes, see the ngram-gpu implementation for an example of how to accomplish this
